File: lucafontanili.designpatterns/src/main/java/com/lucafontanili/designpatterns/creational/prototype/ShapeCache.java
package com.lucafontanili.designpatterns.creational.prototype;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import com.lucafontanili.designpatterns.utilities.Constants;

public class ShapeCache {
    private static final Map<String, AbstractShape> shapeMap = new ConcurrentHashMap<>();

    private ShapeCache() {

    }

    public static void loadCache() {
	shapeMap.put(Constants.CIRCLE, new Circle());
	shapeMap.put(Constants.RECTANGLE, new Rectangle());
    }

    public static AbstractShape getShape(String shapeId) {
	AbstractShape cachedShape = shapeMap.get(shapeId);
	return (AbstractShape) cachedShape.clone();
    }

}

File: lucafontanili.designpatterns.python/creational/prototype/ShapeCache.py
'''
Created on Nov 5, 2016

@author: Admin
'''
from Shape import *

class ShapeCacheClass:
    
    _shape_map = dict()
    
    @staticmethod
    def load_cache():
        ShapeCacheClass._shape_map['circle'] = Circle(1)
        ShapeCacheClass._shape_map['rectangle'] = Rectangle(2)
    
    @staticmethod
    def get_shape(shape_type):
        return ShapeCacheClass._shape_map[shape_type].clone()